<img style="float:right;width:100px;" src="../img/scarecrow.png" alt="scarecrow"/>

[root](root.md)

# deprecated

## scarecrow samtag
The `samtag` tool processes the FASTQ file generated by `reap` and a BAM file generated from aligning that FASTQ file, to add read TAGS {"CR", "CY", "CB", "XP", "XM", "UR", "UY"}. This poses some technical challenges given the large number of reads to cross-reference between the BAM file and the FASTQ file that contains the TAG values. To minimize RAM overhead `scarecrow samtag` creates an sqlite db byte index of the sequence headers in the FASTQ file. After indexing the FASTQ file, the BAM file is split into *n = threads* chunks for parallel processing. Read-specific TAGS are retrieved from the FASTQ by referencing the index and written to a chunk SAM file. Once all chunks are processed the SAM files are combined into the output BAM file and the chunk files removed. This is not memory-intensive (<1 Gb) but can take several hours (6 hrs for 164 M reads using 32 threads) and requires sufficient disk space to complete. Plan for around 16x input BAM file size (i.e. 3 Gb BAM will require ~ 48 Gb headroom).

```bash
scarecrow samtag \
    --threads 16 \
    --fastq extracted.fastq \
    --sam aligned.bam \
    --out tagged.bam
```

The following TAGS are added if found:

| TAG | Description |
| --- | ----------- |
| CR  | Uncorrected barcode sequence at expected position |
| CY  | Uncorrected barcode base qualities |
| CB  | Corrected barcode sequence accounting for jitter |
| XP  | Corrected barcode start position |
| XM  | Corrected barcode mismatch count |
| UR  | UMI sequence |
| UY  | UMI base qualities |

The following is an example of a read with TAGS in the FASTQ sequence header:

```bash
@SRR28867558.17288230 1 VH01123:94:AACNK35M5:1:1203:21356:13097/2 CR=GCTTATAG_ACAACTGT_AGCAGGAA CY=CCCCC-C-_CC---C;-_CCCC---- CB=GCTTATAG_ACCACTGT_AGCAGGAA XP=79_49_11 XM=0_1_0 UR=TACATAACGG UY=;CCC-CCCCC
CGTTCTCCTCAGCACAGACCCGGAGAGCACCGCGAGGGCGGAGCTGCGTACTCCTCTGCACAGATATCGCTGGT
+
;-CC;C-;;CCC-CCCC;C;-C;C-CCCCC-CC;C--;C--CCCC;CC--CCCC-;CC;CCCCC--C;C-;;C-
```

And the equivalent record in the tagged BAM file:

```bash
SRR28867558.17288230    16  hg38_chr1   10537   3   74M *   0   0   ACCAGCGATATCTGTGCAGAGGAGTACGCAGCTCCGCCCTCGCGGTGCTCTCCGGGTCTGTGCTGAGGAGAACG  -C;;-C;C--CCCCC;CC;-CCCC--CC;CCCC--C;--C;CC-CCCCC-C;C-;C;CCCC-CCC;;-C;CC-;  NH:i:2  HI:i:1  AS:i:64 nM:i:4  CR:Z:GCTTATAG_ACAACTGT_AGCAGGAA CY:Z:CCCCC-C-_CC---C;-_CCCC---- CB:Z:GCTTATAG_ACCACTGT_AGCAGGAA XP:Z:79_49_11   XM:Z:0_1_0  UR:Z:TACATAACGG UY:Z:;CCC-CCCCC
```